students/esokada/lesson09/donor_models.py

```diff
@@ -0,0 +1,102 @@
+
+'''
+Donor models for the object-oriented mailroom assignment.
+'''
+
+from operator import itemgetter
+
+
+class Donor():
+    '''
+    Class for individual donors. Includes name, donations, and other information
+    pertaining to a single donor.
+    '''
+
+    def __init__(self, name, donations=None):
+        if isinstance(name, str):
+            self.name = name
+        else:
+            raise TypeError("name must be entered as a string")
+        if donations is None:
+            self.donations = []
+        else:
+            if isinstance(donations, list) is False:
+                raise TypeError("donations must be entered as a list")
+            for i in donations:
+                if isinstance(i, (int, float)) is False:
+                    raise ValueError("donation must be entered as a number")
+            self.donations = donations
+
+    def add_donation(self, donation):
+        '''Add a donation to the donor's list of donations.'''
+        if isinstance(donation, (int, float)) is False:
+            raise ValueError("donation must be entered as a number")
+        self.donations.append(float(donation))
+
+    @property
+    def total_donations(self):
+        '''Return total of donor's donations.'''
+        return sum(self.donations)
+
+    def write_indiv_letter(self):
+        '''Generate a thank-you letter to the donor for their latest donation.'''
+        return f"Dear {self.name}, thank you for your generous donation of " \
+        f"${self.donations[-1]:.2f}."
+
+
+class DonorCollection():
+    '''
+    Class for collection of donors. Includes all operations that require
+    knowledge of multiple donors. The client code accesses individual
+    donors through this collection.
+    '''
+
+    def __init__(self):
+        self.collection = []
+
+    def add_donor(self, name, donations=None):
+        '''Add a new donor to the collection.'''
+        self.collection.append(Donor(name, donations))
+
+    def get_donor(self, name):
+        '''Return the donor if present, otherwise return False.'''
+        for a in self.collection:
+            if a.name == name:
+                return a
+        else:
+            return False
```

Comment:
best to return `None` here since the function isn't expected to return a boolean (best not to mix types)